### PR TITLE
Ds 276 add nl searching keywords to scraper

### PR DESCRIPTION
**Pull request by metanouvelle**
1. automate the process of clean and dedup on given urls.
2. automate the process of applying hyperlink formula within scraper for the purpose of url validating.
3. add definition of B2B for NL

### CODE REVIEW

File: mondu_website_scrapper/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+import sys
+sys.dont_write_bytecode = True
```
Comment (jmunar) [inline mondu_website_scrapper/__init__.py:2]: This looks like a very random statement, some comment about why it's here would help

File: mondu_website_scrapper/gsheet_api/read_from_gsheet.py [hunk 2]
```diff
@@ -1,34 +1,190 @@
...
-    gsheet_client = get_gsheet_client(client_secret=client_secret, scopes=SCOPES)
-    spreadsheet = gsheet_client.open(spreadsheet_name)
-    if worksheet_name in [sheet.title for sheet in spreadsheet.worksheets()]:
-        input_worksheet = spreadsheet.worksheet(worksheet_name)
-        return pd.DataFrame(input_worksheet.get_all_records(), columns=input_columns)
-
-    raise WorksheetNotFound(f"{worksheet_name} not found under the given spreadsheet")
+
+    input_column: str
+    spreadsheet_name: str = None
```
Comment (jmunar) [inline mondu_website_scrapper/gsheet_api/read_from_gsheet.py:37]: dataclass has a way to define default values without using `__post_init__`, but something like:

```python
from dataclasses import Field

class CleanReadFromGsheet:
    ...
    spreadsheet_name: str = Field(default=SPREADSHEET_NAME)
```

File: mondu_website_scrapper/gsheet_api/read_from_gsheet.py [hunk 3]
```diff
@@ -1,34 +1,190 @@
...
-    spreadsheet_name: str = None,
-    worksheet_name: str = "Marketing-Input-Data",
-) -> pd.DataFrame:
+@dataclass
+class CleanReadFromGsheet:
+
     """
     read value from the given gsheet
 
     Returns: a dataframe
```
Comment (jmunar) [inline mondu_website_scrapper/gsheet_api/read_from_gsheet.py:33]: A class doesn't actually return anything, unless you define the `__call__` method. You should describe how to use the class in the docstring

**Review by jmunar — COMMENTED** [review]
Looks good to me, except for some style and documentation issues. Please go through them

File: mondu_website_scrapper/gsheet_api/read_from_gsheet.py [hunk 4]
```diff
@@ -1,34 +1,190 @@
...
+            self.input_worksheet = spreadsheet.worksheet(self.worksheet_name)
+        else:
+            raise WorksheetNotFound(
+                f"{self.worksheet_name} not found under the given spreadsheet"
+            )
+
+        # get all records from the opened gsheet, then filter on the input url column
+        get_value_from_column = []
+        for i in self.input_worksheet.get_all_records():
+            get_value_from_column.append(i[self.url_input_col])
```
Comment (rodrigoazs) [inline mondu_website_scrapper/gsheet_api/read_from_gsheet.py:80]: Use list comprehension:

`get_value_from_column = [i[self.url_input_col] for i in self.input_worksheet.get_all_records()]`

File: mondu_website_scrapper/gsheet_api/read_from_gsheet.py [hunk 5]
```diff
@@ -1,34 +1,190 @@
...
+        self.input_worksheet.update(f"{CLEANED_URL_COLUMN}1", "cleaned_urls")
+        logger.info("cleaned urls are updated to column %s", CLEANED_URL_COLUMN)
+
+        # apply hyperlink formula to the value of cleaned_url_lst,
+        # assign the value to column HYPERLINK_URL_COLUMN,
+        # and then update column HYPERLINK_URL_COLUMN to gsheet.
+
+        for index, value in enumerate(cleaned_url_lst):
+            hyperlink_cells[index].value = (
+                '=HYPERLINK("' + value + '","' + "applied_hyperlink_func" + '")'
```
Comment (rodrigoazs) [inline mondu_website_scrapper/gsheet_api/read_from_gsheet.py:118]: Use f-strings. Pylint and flake8 are not being run on this code?

**Review by rodrigoazs — CHANGES_REQUESTED** [review]

File: mondu_website_scrapper/spiders/catch_fish_scraper.py [hunk 6]
```diff
@@ -200,6 +205,21 @@ def extract_b2b_keywords(self, response: scrapy.http.response) -> list:
             b2b_key_lst += [word] if re.search(word, data) is not None else []
         return b2b_key_lst
 
+    def _extract_thuiswinkel_certificate(self, response: scrapy.http.response) -> list:
+        """
+        extract bb2 keywords pre-defined under settings.py from html body
+
+        Returns: a list of b2b keywords if they are presented
+        """
+        keyword = "thuiswinke"
```
Comment (jmunar) [inline mondu_website_scrapper/spiders/catch_fish_scraper.py:214]: Is it `thuiswinke` and not `thuiswinkel` on purpose?

File: mondu_website_scrapper/spiders/catch_fish_scraper.py [hunk 7]
```diff
@@ -200,6 +205,21 @@ def extract_b2b_keywords(self, response: scrapy.http.response) -> list:
             b2b_key_lst += [word] if re.search(word, data) is not None else []
         return b2b_key_lst
 
+    def _extract_thuiswinkel_certificate(self, response: scrapy.http.response) -> list:
```
Comment (jmunar) [inline mondu_website_scrapper/spiders/catch_fish_scraper.py:208]: It is bad architecture to have a configuration file setting the parameters for each country, and at the same time having country-specific logic here. The usual solution is to subclass for each country, adding the specific bits. I think it's fine for now, but I would recommend to add decoupling this to the webscraper roadmap

Comment (metanouvelle) [reply mondu_website_scrapper/spiders/catch_fish_scraper.py:208]: let us discuss it through a call

File: mondu_website_scrapper/spiders/catch_fish_scraper.py (hunk 6 above)
Comment (metanouvelle) [reply mondu_website_scrapper/spiders/catch_fish_scraper.py:214]: typo

Comment (metanouvelle) [reply mondu_website_scrapper/spiders/catch_fish_scraper.py]: fixed

**Review by jmunar — APPROVED** [review]

File: mondu_website_scrapper/gsheet_api/read_from_gsheet.py (hunk 4 above)
Comment (metanouvelle) [reply mondu_website_scrapper/gsheet_api/read_from_gsheet.py]: fixed

File: mondu_website_scrapper/gsheet_api/read_from_gsheet.py (hunk 5 above)
Comment (metanouvelle) [reply mondu_website_scrapper/gsheet_api/read_from_gsheet.py]: fixed

**Pull request merged**